File: simplified-tests/src/main/java/org/nypl/simplified/tests/books/accounts/AccountsDatabaseContract.java
package org.nypl.simplified.tests.books.accounts;

import com.io7m.jfunctional.Option;

import org.hamcrest.BaseMatcher;
import org.hamcrest.Description;
import org.junit.Assert;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;
import org.nypl.simplified.books.accounts.AccountAuthenticationCredentials;
import org.nypl.simplified.books.accounts.AccountBarcode;
import org.nypl.simplified.books.accounts.AccountPIN;
import org.nypl.simplified.books.accounts.AccountProvider;
import org.nypl.simplified.books.accounts.AccountProviderCollection;
import org.nypl.simplified.books.accounts.AccountProviderCollectionType;
import org.nypl.simplified.books.accounts.AccountType;
import org.nypl.simplified.books.accounts.AccountsDatabase;
import org.nypl.simplified.books.accounts.AccountsDatabaseException;
import org.nypl.simplified.books.accounts.AccountsDatabaseType;
import org.nypl.simplified.books.book_database.BookDatabaseFactoryType;
import org.nypl.simplified.books.book_database.BookDatabases;
import org.nypl.simplified.books.core.LogUtilities;
import org.nypl.simplified.books.profiles.ProfileID;
import org.nypl.simplified.files.DirectoryUtilities;
import org.nypl.simplified.files.FileUtilities;
import org.slf4j.Logger;

import java.io.File;
import java.io.IOException;
import java.net.URI;
import java.util.SortedMap;
import java.util.TreeMap;

public abstract class AccountsDatabaseContract {

  private static final Logger LOG = LogUtilities.getLog(AccountsDatabaseContract.class);

  @Rule public ExpectedException expected = ExpectedException.none();

  /**
   * An exception matcher that checks to see if the given accounts database exception has
   * at least one cause of the given type and with the given exception message.
   *
   * @param <T> The cause type
   */

  private static final class CausesContains<T extends Exception>
      extends BaseMatcher<AccountsDatabaseException> {
    private final Class<T> exception_type;
    private final String message;

    CausesContains(
        final Class<T> exception_type,
        final String message) {
      this.exception_type = exception_type;
      this.message = message;
    }

    @Override
    public boolean matches(final Object item) {
      if (item instanceof AccountsDatabaseException) {
        final AccountsDatabaseException ex = (AccountsDatabaseException) item;
        for (final Exception c : ex.causes()) {
          LOG.error("Cause: ", c);
          if (exception_type.isAssignableFrom(c.getClass()) && c.getMessage().contains(message)) {
            return true;
          }
        }
      }
      return false;
    }

    @Override
    public void describeTo(final Description description) {
      description.appendText("must throw AccountsDatabaseException");
      description.appendText(" with at least one cause of type " + exception_type);
      description.appendText(" with a message containing '" + message + "'");
    }
  }

  @Test
  public final void testOpenExistingNotDirectory()
      throws Exception {

    final File f_tmp = DirectoryUtilities.directoryCreateTemporary();
    final File f_pro = new File(f_tmp, "profiles");
    f_pro.mkdirs();
    final File f_p = new File(f_pro, "0");
    f_p.mkdirs();
    final File f_acc = new File(f_p, "accounts");

    FileUtilities.fileWriteUTF8(f_acc, "Hello!");

    expected.expect(AccountsDatabaseException.class);
    expected.expect(new CausesContains<>(IOException.class, "Not a directory"));
    AccountsDatabase.open(bookDatabases(), accountProviders(), f_acc);
  }

  private BookDatabases bookDatabases() {
    return BookDatabases.get();
  }

  @Test
  public final void testOpenExistingBadSubdirectory()
      throws Exception {

    final File f_tmp = DirectoryUtilities.directoryCreateTemporary();
    final File f_pro = new File(f_tmp, "profiles");
    f_pro.mkdirs();
    final File f_p = new File(f_pro, "0");
    f_p.mkdirs();
    final File f_acc = new File(f_p, "accounts");
    f_acc.mkdirs();
    final File f_a = new File(f_acc, "xyz");
    f_a.mkdirs();

    expected.expect(AccountsDatabaseException.class);
    expected.expect(new CausesContains<>(
        IOException.class, "Could not parse directory name as an account ID"));
    AccountsDatabase.open(bookDatabases(), accountProviders(), f_acc);
  }

  @Test
  public final void testOpenExistingJSONMissing()
      throws Exception {

    final File f_tmp = DirectoryUtilities.directoryCreateTemporary();
    final File f_pro = new File(f_tmp, "profiles");
    f_pro.mkdirs();
    final File f_p = new File(f_pro, "0");
    f_p.mkdirs();
    final File f_acc = new File(f_p, "accounts");
    f_acc.mkdirs();
    final File f_a = new File(f_acc, "0");
    f_a.mkdirs();

    expected.expect(AccountsDatabaseException.class);
    expected.expect(new CausesContains<>(IOException.class, "Could not parse account: "));
    AccountsDatabase.open(bookDatabases(), accountProviders(), f_acc);
  }

  @Test
  public final void testOpenExistingJSONUnparseable()
      throws Exception {

    final File f_tmp = DirectoryUtilities.directoryCreateTemporary();
    final File f_pro = new File(f_tmp, "profiles");
    f_pro.mkdirs();
    final File f_p = new File(f_pro, "0");
    f_p.mkdirs();
    final File f_acc = new File(f_p, "accounts");
    f_acc.mkdirs();
    final File f_a = new File(f_acc, "0");
    f_a.mkdirs();

    final File f_f = new File(f_a, "account.json");
    FileUtilities.fileWriteUTF8(f_f, "} { this is not JSON { } { }");

    expected.expect(AccountsDatabaseException.class);
    expected.expect(new CausesContains<>(IOException.class, "Could not parse account: "));
    AccountsDatabase.open(bookDatabases(), accountProviders(), f_acc);
  }

  @Test
  public final void testOpenExistingEmpty()
      throws Exception {

    final File f_tmp = DirectoryUtilities.directoryCreateTemporary();
    final File f_pro = new File(f_tmp, "profiles");
    f_pro.mkdirs();
    final File f_p = new File(f_pro, "0");
    f_p.mkdirs();
    final File f_acc = new File(f_p, "accounts");

    final AccountsDatabaseType db = 
        AccountsDatabase.open(bookDatabases(), accountProviders(), f_acc);
    Assert.assertEquals(0, db.accounts().size());
    Assert.assertEquals(f_acc, db.directory());
  }

  @Test
  public final void testCreateAccount()
      throws Exception {

    final File f_tmp = DirectoryUtilities.directoryCreateTemporary();
    final File f_pro = new File(f_tmp, "profiles");
    f_pro.mkdirs();
    final File f_p = new File(f_pro, "0");
    f_p.mkdirs();
    final File f_acc = new File(f_p, "accounts");

    final AccountProviderCollectionType account_providers =
        accountProviders();
    final AccountsDatabaseType db =
        AccountsDatabase.open(bookDatabases(), account_providers, f_acc);

    final AccountProvider provider0 =
        fakeProvider("http://www.example.com/accounts0/");
    final AccountProvider provider1 =
        fakeProvider("http://www.example.com/accounts1/");

    final AccountType acc0 = db.createAccount(provider0);
    final AccountType acc1 = db.createAccount(provider1);

    Assert.assertTrue("Account 0 directory exists", acc0.directory().isDirectory());
    Assert.assertTrue("Account 1 directory exists", acc1.directory().isDirectory());

    Assert.assertTrue(
        "Account 0 file exists",
        new File(acc0.directory(), "account.json").isFile());
    Assert.assertTrue(
        "Account 1 file exists",
        new File(acc1.directory(), "account.json").isFile());

    Assert.assertEquals(
        account_providers.provider(URI.create("http://www.example.com/accounts0/")),
        acc0.provider());
    Assert.assertEquals(
        account_providers.provider(URI.create("http://www.example.com/accounts1/")),
        acc1.provider());

    Assert.assertNotEquals(acc0.id(), acc1.id());
    Assert.assertNotEquals(acc0.directory(), acc1.directory());

    Assert.assertEquals(Option.none(), acc0.credentials());
    Assert.assertEquals(Option.none(), acc1.credentials());
  }

  @Test
  public final void testCreateReopen()
      throws Exception {

    final File f_tmp = DirectoryUtilities.directoryCreateTemporary();
    final File f_pro = new File(f_tmp, "profiles");
    f_pro.mkdirs();
    final File f_p = new File(f_pro, "0");
    f_p.mkdirs();
    final File f_acc = new File(f_p, "accounts");

    final AccountsDatabaseType db0 = 
        AccountsDatabase.open(bookDatabases(), accountProviders(), f_acc);

    final AccountProvider provider0 =
        fakeProvider("http://www.example.com/accounts0/");
    final AccountProvider provider1 =
        fakeProvider("http://www.example.com/accounts1/");

    final AccountType acc0 = db0.createAccount(provider0);
    final AccountType acc1 = db0.createAccount(provider1);

    final AccountsDatabaseType db1 = 
        AccountsDatabase.open(bookDatabases(), accountProviders(), f_acc);

    final AccountType acr0 = db1.accounts().get(acc0.id());
    final AccountType acr1 = db1.accounts().get(acc1.id());

    Assert.assertEquals(acc0.id(), acr0.id());
    Assert.assertEquals(acc1.id(), acr1.id());
    Assert.assertEquals(acc0.directory(), acr0.directory());
    Assert.assertEquals(acc1.directory(), acr1.directory());
    Assert.assertEquals(acc0.provider(), acr0.provider());
    Assert.assertEquals(acc1.provider(), acr1.provider());
  }

  @Test
  public final void testSetCredentials()
      throws Exception {

    final File f_tmp = DirectoryUtilities.directoryCreateTemporary();
    final File f_pro = new File(f_tmp, "profiles");
    f_pro.mkdirs();
    final File f_p = new File(f_pro, "0");
    f_p.mkdirs();
    final File f_acc = new File(f_p, "accounts");

    final AccountsDatabaseType db0 = AccountsDatabase.open(bookDatabases(), accountProviders(), f_acc);
    final AccountProvider provider0 = fakeProvider("http://www.example.com/accounts0/");
    final AccountType acc0 = db0.createAccount(provider0);

    final AccountAuthenticationCredentials creds =
        AccountAuthenticationCredentials.builder(
            AccountPIN.create("1234"),
            AccountBarcode.create("1234"))
            .build();

    acc0.setCredentials(Option.some(creds));
    Assert.assertEquals(Option.some(creds), acc0.credentials());
  }

  private AccountProviderCollectionType accountProviders() {
    final AccountProvider p0 = fakeProvider("http://www.example.com/accounts0/");
    final AccountProvider p1 = fakeProvider("http://www.example.com/accounts1/");
    final SortedMap<URI, AccountProvider> providers = new TreeMap<>();
    providers.put(p0.id(), p0);
    providers.put(p1.id(), p1);
    return AccountProviderCollection.create(p0, providers);
  }

  private static AccountProvider fakeProvider(final String provider_id) {
    return AccountProvider.builder()
        .setId(URI.create(provider_id))
        .setDisplayName("Fake Library")
        .setSubtitle("Imaginary books")
        .setLogo(URI.create("http://example.com/logo.png"))
        .setCatalogURI(URI.create("http://example.com/accounts0/feed.xml"))
        .setSupportEmail("postmaster@example.com")
        .build();
  }
}

File: simplified-tests/src/test/java/org/nypl/simplified/tests/books/accounts/AccountsDatabaseContract.kt
package org.nypl.simplified.tests.books.accounts

import android.content.Context
import io.reactivex.subjects.PublishSubject
import org.hamcrest.BaseMatcher
import org.hamcrest.Description
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import org.nypl.simplified.accounts.api.AccountAuthenticationCredentials
import org.nypl.simplified.accounts.api.AccountEvent
import org.nypl.simplified.accounts.api.AccountLoginState
import org.nypl.simplified.accounts.api.AccountPassword
import org.nypl.simplified.accounts.api.AccountUsername
import org.nypl.simplified.accounts.database.AccountsDatabase
import org.nypl.simplified.accounts.database.api.AccountsDatabaseDuplicateProviderException
import org.nypl.simplified.accounts.database.api.AccountsDatabaseException
import org.nypl.simplified.accounts.registry.api.AccountProviderRegistryType
import org.nypl.simplified.books.book_database.BookDatabases
import org.nypl.simplified.files.DirectoryUtilities
import org.nypl.simplified.files.FileUtilities
import org.nypl.simplified.profiles.api.ProfileEvent
import org.nypl.simplified.tests.mocking.FakeAccountCredentialStorage
import org.nypl.simplified.tests.mocking.MockAccountProviders
import org.slf4j.LoggerFactory
import java.io.File
import java.io.IOException
import java.net.URI

abstract class AccountsDatabaseContract {

  private val logger = LoggerFactory.getLogger(AccountsDatabaseContract::class.java)

  private lateinit var accountEvents: PublishSubject<AccountEvent>
  private lateinit var accountProviders: AccountProviderRegistryType
  private lateinit var credentialStore: FakeAccountCredentialStorage
  private lateinit var profileEvents: PublishSubject<ProfileEvent>

  protected abstract fun context(): Context

  @BeforeEach
  open fun setup() {
    this.credentialStore = FakeAccountCredentialStorage()
    this.accountEvents = PublishSubject.create()
    this.profileEvents = PublishSubject.create()
    this.accountProviders = Mockito.mock(AccountProviderRegistryType::class.java)
  }

  /**
   * An exception matcher that checks to see if the given accounts database exception has
   * at least one cause of the given type and with the given exception message.
   *
   * @param <T> The cause type
   */

  private class CausesContains<T : Exception>(
    private val exception_type: Class<T>,
    private val message: String
  ) : BaseMatcher<AccountsDatabaseException>() {

    private val logger = LoggerFactory.getLogger(CausesContains::class.java)

    override fun matches(item: Any): Boolean {
      if (item is AccountsDatabaseException) {
        for (c in item.causes()) {
          this.logger.error("Cause: ", c)
          if (this.exception_type.isAssignableFrom(c.javaClass) && c.message!!.contains(this.message)) {
            return true
          }
        }
      }
      return false
    }

    override fun describeTo(description: Description) {
      description.appendText("must throw AccountsDatabaseException")
      description.appendText(" with at least one cause of type ${this.exception_type}")
      description.appendText(" with a message containing '${this.message}'")
    }
  }

  @Test
  @Throws(Exception::class)
  fun testOpenExistingNotDirectory() {
    val fileTemp = DirectoryUtilities.directoryCreateTemporary()
    val fileProfiles = File(fileTemp, "profiles")
    fileProfiles.mkdirs()
    val f_p = File(fileProfiles, "0")
    f_p.mkdirs()
    val f_acc = File(f_p, "accounts")

    FileUtilities.fileWriteUTF8(f_acc, "Hello!")

    val ex = Assertions.assertThrows(AccountsDatabaseException::class.java) {
      AccountsDatabase.open(
        this.context(),
        this.accountEvents,
        this.bookDatabases(),
        this.credentialStore,
        this.accountProviders,
        f_acc
      )
    }

    val causes =
      ex.causes()
        .filterIsInstance<IOException>()
        .find { ex -> ex.message!!.contains("Not a directory") }
    Assertions.assertTrue(causes != null)
  }

  private fun bookDatabases(): BookDatabases {
    return BookDatabases
  }

  private fun onAccountResolutionStatus(
    accountProvider: URI,
    message: String
  ) {
    this.logger.debug("onAccountResolutionStatus: {}: {}", accountProvider, message)
  }

  /**
   * A bad subdirectory will be migrated but will still fail to open.
   */

  @Test
  @Throws(Exception::class)
  fun testOpenExistingBadSubdirectory() {
    val fileTemp = DirectoryUtilities.directoryCreateTemporary()
    val fileProfiles = File(fileTemp, "profiles")
    fileProfiles.mkdirs()
    val f_p = File(fileProfiles, "0")
    f_p.mkdirs()
    val f_acc = File(f_p, "accounts")
    f_acc.mkdirs()
    val f_a = File(f_acc, "xyz")
    f_a.mkdirs()

    val ex = Assertions.assertThrows(AccountsDatabaseException::class.java) {
      val database = AccountsDatabase.open(
        this.context(),
        this.accountEvents,
        this.bookDatabases(),
        this.credentialStore,
        this.accountProviders,
        f_acc
      )
    }

    val causes =
      ex.causes()
        .filterIsInstance<IOException>()
        .find { ex -> ex.message!!.contains("Could not parse account: ") }
    Assertions.assertTrue(causes != null)
  }

  @Test
  @Throws(Exception::class)
  fun testOpenExistingJSONMissing() {
    val fileTemp = DirectoryUtilities.directoryCreateTemporary()
    val fileProfiles = File(fileTemp, "profiles")
    fileProfiles.mkdirs()
    val f_p = File(fileProfiles, "0")
    f_p.mkdirs()
    val f_acc = File(f_p, "accounts")
    f_acc.mkdirs()
    val f_a = File(f_acc, "64e606e8-e242-4c5f-9bfb-63df11b187bd")
    f_a.mkdirs()

    val ex = Assertions.assertThrows(AccountsDatabaseException::class.java) {
      AccountsDatabase.open(
        this.context(),
        this.accountEvents,
        this.bookDatabases(),
        this.credentialStore,
        this.accountProviders,
        f_acc
      )
    }

    val causes =
      ex.causes()
        .filterIsInstance<IOException>()
        .find { ex -> ex.message!!.contains("Could not parse account: ") }
    Assertions.assertTrue(causes != null)
  }

  @Test
  @Throws(Exception::class)
  fun testOpenExistingJSONUnparseable() {
    val fileTemp = DirectoryUtilities.directoryCreateTemporary()
    val fileProfiles = File(fileTemp, "profiles")
    fileProfiles.mkdirs()
    val f_p = File(fileProfiles, "0")
    f_p.mkdirs()
    val f_acc = File(f_p, "accounts")
    f_acc.mkdirs()
    val f_a = File(f_acc, "0")
    f_a.mkdirs()

    val f_f = File(f_a, "account.json")
    FileUtilities.fileWriteUTF8(f_f, "} { this is not JSON { } { }")

    val ex = Assertions.assertThrows(AccountsDatabaseException::class.java) {
      AccountsDatabase.open(
        this.context(),
        this.accountEvents,
        this.bookDatabases(),
        this.credentialStore,
        this.accountProviders,
        f_acc
      )
    }

    val causes =
      ex.causes()
        .filterIsInstance<IOException>()
        .find { ex -> ex.message!!.contains("Could not parse account: ") }
    Assertions.assertTrue(causes != null)
  }

  @Test
  @Throws(Exception::class)
  fun testOpenExistingEmpty() {
    val fileTemp = DirectoryUtilities.directoryCreateTemporary()
    val fileProfiles = File(fileTemp, "profiles")
    fileProfiles.mkdirs()
    val f_p = File(fileProfiles, "0")
    f_p.mkdirs()
    val f_acc = File(f_p, "accounts")

    val db = AccountsDatabase.open(
      this.context(),
      this.accountEvents,
      this.bookDatabases(),
      this.credentialStore,
      this.accountProviders,
      f_acc
    )

    Assertions.assertEquals(0, db.accounts().size.toLong())
    Assertions.assertEquals(f_acc, db.directory())
  }

  @Test
  @Throws(Exception::class)
  fun testCreateAccount() {
    val fileTemp = DirectoryUtilities.directoryCreateTemporary()
    val fileProfiles = File(fileTemp, "profiles")
    fileProfiles.mkdirs()
    val f_p = File(fileProfiles, "0")
    f_p.mkdirs()
    val f_acc = File(f_p, "accounts")

    val accountProviders =
      MockAccountProviders.fakeAccountProviders()

    val db = AccountsDatabase.open(
      this.context(),
      this.accountEvents,
      this.bookDatabases(),
      this.credentialStore,
      this.accountProviders,
      f_acc
    )

    val provider0 =
      MockAccountProviders.fakeProvider("http://www.example.com/accounts0/")
    val provider1 =
      MockAccountProviders.fakeProvider("http://www.example.com/accounts1/")

    val acc0 = db.createAccount(provider0)
    val acc1 = db.createAccount(provider1)

    Assertions.assertTrue(acc0.directory.isDirectory, "Account 0 directory exists")
    Assertions.assertTrue(acc1.directory.isDirectory, "Account 1 directory exists")
    Assertions.assertTrue(File(acc0.directory, "account.json").isFile, "Account 0 file exists")
    Assertions.assertTrue(File(acc1.directory, "account.json").isFile, "Account 1 file exists")

    Assertions.assertEquals(provider0, acc0.provider)
    Assertions.assertEquals(provider1, acc1.provider)

    Assertions.assertNotEquals(acc0.id, acc1.id)
    Assertions.assertNotEquals(acc0.directory, acc1.directory)

    Assertions.assertEquals(AccountLoginState.AccountNotLoggedIn, acc0.loginState)
    Assertions.assertEquals(AccountLoginState.AccountNotLoggedIn, acc1.loginState)
  }

  @Test
  @Throws(Exception::class)
  fun testCreateAccountProviderAlreadyUsed() {
    val fileTemp = DirectoryUtilities.directoryCreateTemporary()
    val fileProfiles = File(fileTemp, "profiles")
    fileProfiles.mkdirs()
    val f_p = File(fileProfiles, "0")
    f_p.mkdirs()
    val f_acc = File(f_p, "accounts")

    val db =
      AccountsDatabase.open(
        this.context(),
        this.accountEvents,
        this.bookDatabases(),
        this.credentialStore,
        this.accountProviders,
        f_acc
      )

    val provider0 =
      MockAccountProviders.fakeProvider("http://www.example.com/accounts0/")

    val acc0 = db.createAccount(provider0)

    Assertions.assertThrows(AccountsDatabaseDuplicateProviderException::class.java) {
      val acc1 = db.createAccount(provider0)
    }
  }

  @Test
  @Throws(Exception::class)
  fun testCreateReopen() {
    val fileTemp = DirectoryUtilities.directoryCreateTemporary()
    val fileProfiles = File(fileTemp, "profiles")
    fileProfiles.mkdirs()
    val f_p = File(fileProfiles, "0")
    f_p.mkdirs()
    val f_acc = File(f_p, "accounts")

    val db0 = AccountsDatabase.open(
      this.context(),
      this.accountEvents,
      this.bookDatabases(),
      this.credentialStore,
      this.accountProviders,
      f_acc
    )

    val provider0 =
      MockAccountProviders.fakeProvider("urn:fake:0")
    val provider1 =
      MockAccountProviders.fakeProvider("urn:fake:1")

    val acc0 = db0.createAccount(provider0)
    val acc1 = db0.createAccount(provider1)

    val db1 = AccountsDatabase.open(
      this.context(),
      this.accountEvents,
      this.bookDatabases(),
      this.credentialStore,
      this.accountProviders,
      f_acc
    )

    val acr0 = db1.accounts()[acc0.id]!!
    val acr1 = db1.accounts()[acc1.id]!!

    Assertions.assertEquals(acc0.id, acr0.id)
    Assertions.assertEquals(acc1.id, acr1.id)
    Assertions.assertEquals(acc0.directory, acr0.directory)
    Assertions.assertEquals(acc1.directory, acr1.directory)
    Assertions.assertEquals(acc0.provider.id, acr0.provider.id)
    Assertions.assertEquals(acc1.provider.id, acr1.provider.id)
  }

  @Test
  @Throws(Exception::class)
  fun testSetCredentials() {
    val fileTemp = DirectoryUtilities.directoryCreateTemporary()
    val fileProfiles = File(fileTemp, "profiles")
    fileProfiles.mkdirs()
    val f_p = File(fileProfiles, "0")
    f_p.mkdirs()
    val f_acc = File(f_p, "accounts")

    val db0 = AccountsDatabase.open(
      this.context(),
      this.accountEvents,
      this.bookDatabases(),
      this.credentialStore,
      this.accountProviders,
      f_acc
    )

    val provider0 =
      MockAccountProviders.fakeProvider("http://www.example.com/accounts0/")
    val acc0 = db0.createAccount(provider0)

    val creds =
      AccountAuthenticationCredentials.Basic(
        userName = AccountUsername("abcd"),
        password = AccountPassword("1234"),
        adobeCredentials = null,
        authenticationDescription = null,
        annotationsURI = URI("https://www.example.com")
      )

    acc0.setLoginState(AccountLoginState.AccountLoggedIn(creds))
    Assertions.assertEquals(AccountLoginState.AccountLoggedIn(creds), acc0.loginState)

    acc0.updateCredentialsIfAvailable {
      when (it) {
        is AccountAuthenticationCredentials.Basic ->
          creds.copy(annotationsURI = URI.create("https://www.example.com/annotations"))
        is AccountAuthenticationCredentials.OAuthWithIntermediary ->
          creds.copy(annotationsURI = URI.create("https://www.example.com/annotations"))
        is AccountAuthenticationCredentials.SAML2_0 ->
          creds.copy(annotationsURI = URI.create("https://www.example.com/annotations"))
      }
    }

    Assertions.assertEquals(
      URI.create("https://www.example.com/annotations"),
      acc0.loginState.credentials?.annotationsURI
    )
  }

  @Test
  @Throws(Exception::class)
  fun testSetProviderWrongID() {
    val fileTemp = DirectoryUtilities.directoryCreateTemporary()
    val fileProfiles = File(fileTemp, "profiles")
    fileProfiles.mkdirs()
    val f_p = File(fileProfiles, "0")
    f_p.mkdirs()
    val f_acc = File(f_p, "accounts")

    val db0 = AccountsDatabase.open(
      this.context(),
      this.accountEvents,
      this.bookDatabases(),
      this.credentialStore,
      this.accountProviders,
      f_acc
    )

    val provider0 =
      MockAccountProviders.fakeProvider("http://www.example.com/accounts0/")
    val provider1 =
      MockAccountProviders.fakeProvider("http://www.example.com/accounts1/")

    val acc0 = db0.createAccount(provider0)

    Assertions.assertThrows(AccountsDatabaseException::class.java) {
      acc0.setAccountProvider(provider1)
    }
  }

  @Test
  @Throws(Exception::class)
  fun testSetProviderOK() {
    val fileTemp = DirectoryUtilities.directoryCreateTemporary()
    val fileProfiles = File(fileTemp, "profiles")
    fileProfiles.mkdirs()
    val f_p = File(fileProfiles, "0")
    f_p.mkdirs()
    val f_acc = File(f_p, "accounts")

    val db0 = AccountsDatabase.open(
      this.context(),
      this.accountEvents,
      this.bookDatabases(),
      this.credentialStore,
      this.accountProviders,
      f_acc
    )

    val provider0 =
      MockAccountProviders.fakeProvider("http://www.example.com/accounts0/")
    val provider1 =
      MockAccountProviders.fakeProvider("http://www.example.com/accounts0/")

    val acc0 = db0.createAccount(provider0)
    acc0.setAccountProvider(provider1)
  }
}
